src/analysis/lattice.h

```diff
@@ -59,9 +59,25 @@ concept Lattice = requires(const L& lattice,
   { lattice.join(elem, constElem) } noexcept -> std::same_as<bool>;
 };
 
+// The analysis framework only uses bottom elements and least upper bounds (i.e.
+// joins) directly, so lattices do not necessarily need to implement top
+// elements and greatest lower bounds (i.e. meets) to be useable, even though
+// they are required for mathematical lattices. Implementing top elements and
+// meets does have the benefit of making a lattice generically invertable,
+// though. See lattices/inverted.h.
+template<typename L>
+concept FullLattice =
+  Lattice<L> && requires(const L& lattice,
+                         const typename L::Element& constElem,
```

Comment:
Can you explain why access to L::Element& is needed as both a const and !const?

Reply:
This is for the `join` and `meet` operations, which need to take a mutable reference as their first argument because it will be modified in-place. Otherwise all other uses of elements are `const`.